File: problems-adapt/0101-0200/0139_spelling-the-string-from-the-dictionary/solution_trie_walk.go
func canSpellFromDictionary(s string, dictionary []string) bool {
	// Trie over the dictionary: nodes own a child map keyed by letter plus
	// the flag marking a node where a word ends. From every reachable
	// position a walk follows s's own characters, so a branch dies at the
	// first character no remaining word shares, and each terminal crossed
	// marks the prefix after it reachable.
	type trieNode struct {
		children map[byte]*trieNode
		end      bool
	}
	root := &trieNode{children: map[byte]*trieNode{}}
	for _, word := range dictionary {
		node := root
		for i := 0; i < len(word); i++ {
			child, ok := node.children[word[i]]
			if !ok {
				child = &trieNode{children: map[byte]*trieNode{}}
				node.children[word[i]] = child
			}
			node = child
		}
		node.end = true
	}
	n := len(s)
	reachable := make([]bool, n+1)
	reachable[0] = true
	for i := 0; i < n; i++ {
		if !reachable[i] {
			continue
		}
		node := root
		for j := i; j < n; j++ {
			child, ok := node.children[s[j]]
			if !ok {
				break
			}
			node = child
			// Every terminal on the path ends a word at this depth.
			if node.end {
				reachable[j+1] = true
			}
		}
	}
	return reachable[n]
}

File: problems-adapt/0101-0200/0139_spelling-the-string-from-the-dictionary/solution_trie_walk.rs
use std::collections::HashMap;

impl Solution {
    pub fn can_spell_from_dictionary(s: String, dictionary: Vec<String>) -> bool {
        // Trie over the dictionary: nodes own a child map keyed by letter
        // plus the flag marking a node where a word ends. From every
        // reachable position a walk follows s's own characters, so a branch
        // dies at the first character no remaining word shares, and each
        // terminal crossed marks the prefix after it reachable.
        #[derive(Default)]
        struct TrieNode {
            children: HashMap<u8, TrieNode>,
            end: bool,
        }
        let mut root = TrieNode::default();
        for word in &dictionary {
            let mut node = &mut root;
            for ch in word.as_bytes() {
                node = node.children.entry(*ch).or_default();
            }
            node.end = true;
        }
        let n = s.len();
        let bytes = s.as_bytes();
        let mut reachable = vec![false; n + 1];
        reachable[0] = true;
        for i in 0..n {
            if !reachable[i] {
                continue;
            }
            let mut node = &root;
            for j in i..n {
                match node.children.get(&bytes[j]) {
                    None => break,
                    Some(child) => {
                        node = child;
                        // Every terminal on the path ends a word at this depth.
                        if node.end {
                            reachable[j + 1] = true;
                        }
                    }
                }
            }
        }
        reachable[n]
    }
}
